pkg/dockerregistry/server/writelimiter/semaphore.go

```diff
@@ -0,0 +1,24 @@
+package writelimiter
+
+type semaphore struct {
+	c chan struct{}
+}
+
+func newSemaphore(count int) *semaphore {
+	return &semaphore{
+		c: make(chan struct{}, count),
+	}
+}
+
+func (s *semaphore) Up() {
+	<-s.c
+}
+
+func (s *semaphore) TryDown() bool {
```

Comment:
> it would be better for this to be structured in a way that cannot fail

We can't structure it this way, because our middleware can't control objects lifetime.
[Create](https://github.com/docker/distribution/blob/91c507a39abfce14b5c8541cf284330e22208c0f/registry/handlers/blobupload.go#L121) and [Resume](https://github.com/docker/distribution/blob/91c507a39abfce14b5c8541cf284330e22208c0f/registry/handlers/blobupload.go#L63) are called from the upstream code. There are already exist Commit/Close/Cancel operations, but sometimes neither of them is called ([1], [2]).

> probably 30s or 1m
Is it how long the request should wait in the queue? Because context deadlines are normally used to limit how long an operation can proceed. And 30 seconds might be not enough to accept 1 GB layer.

[1]: https://github.com/openshift/origin/blob/198b66f419b06e76a67fb67a7c74820d433e49cf/pkg/dockerregistry/server/pullthroughblobstore.go#L220-L223
[2]: https://github.com/docker/distribution/blob/91c507a39abfce14b5c8541cf284330e22208c0f/registry/proxy/proxyblobstore.go#L105